### ISSUE TITLE
autoconf tools issue on Windows/MinGW-w64

### ISSUE THREAD

**Issue by brechtsanders** [llvm/llvm-project#59873]
I just figured out why autoconf tools fail to build shared libraries when using clang on native Windows (MinGW-w64 target).

Apparently `configure` checks for the result of `$LD --help 2>&1 | $GREP 'auto-import'`.

When `LD` is set to binutils' `ld` there some matches, including `--enable-auto-import` and `--disable-auto-import`.

But when `LD` is set to LLVM's `ld.lld` there are no matches at all.

This leads to `configure` scripts never generating shared libraries on Windows.

This dirty workaround allows me to build shared libraries for a large number of projects when using LLVM:
```sh
sed -i.bak -e "s/\$LD --help 2>&1 | \$GREP 'auto-import'/true/" configure
```

Would you consider adding a line for the `--enable-auto-import` flag in the help show when running `ld.lld --help`, so autoconf tools can pick it up and goes ahead to build shard libraries?

The flag itself can be a dummy, though I think it's already accepted by the linker.

**Comment by mstorsjo** [comment]
Hi @brechtsanders!

Actually, lld does print `--enable-auto-import` in the lld help output - if invoking the right lld interface (the mingw one).

`ld.lld` is generally speaking the interface to the ELF linker, so if you run `ld.lld --help`, you get the help for the ELF part of lld. If you instead run e.g. `ld.lld -m i386pep --help`, you get the help for the MinGW linker interface, which does include the text that libtool looks for.

(Side note; see 6ff6a8f656eeb59f2051568334b82e1615bdca48 - where we added this option to be printed in the help listing in 2019, even if the option itself actually didn't do anything at the time. In 2020, in 7f0e6c31c25530352aa6e4cd9e2bdc4381368eba, the option was changed to actually have an effect, overriding an earlier `--disable-auto-import`.)

Normally when the compiler invokes the linker, it does pass the right `-m` options, so the issue is primarily with use cases like these, that invokes `ld` directly to check the output of e.g. `--help` and similar.

In llvm-mingw, this is worked around by having `ld` be a shell script which wraps `ld.lld` and always calls it with a correct `-m` flag: https://github.com/mstorsjo/llvm-mingw/blob/master/wrappers/ld-wrapper.sh#L28-L34 This has been in place since the inception of llvm-mingw. However later, msys2 also ran into the same issue, and in 61e0b2b4c5fbbea01bb40f28ea0222b87166ccdf a lld CMake build option was added, allowing making `ld.lld` default to a mingw target instead of an ELF target, if no `-m` option is specified.

So TL;DR - set `-DLLD_DEFAULT_LD_LLD_IS_MINGW=TRUE` when you build lld, and this issue should go away.

**Comment by brechtsanders** [comment]
Hi @mstorsjo!
Looks like again I came across something you were already quite experienced with ;-)
Building LLVM with `-DLLD_DEFAULT_LD_LLD_IS_MINGW=TRUE` does fix this issue. Thanks!

To be honest the llvm-mingw wrapper script trick seems like a bit of a dirty trick to me, but it actually works quite well.
I am now trying to build a lot of libraries without wrapper scripts by just setting environment variables like `CC`/`CXX`/`LD`.

Your answer here already removes the first hurdle.
Thanks!

**Comment by mstorsjo** [comment]
> To be honest the llvm-mingw wrapper script trick seems like a bit of a dirty trick to me, but it actually works quite well.

Yep, it's not great, but it serves the purpose of both switching lld to mingw mode, and for inferring the architecture from the prefix.

For some tools in llvm (llvm-windres and llvm-dlltool) the tool itself can already interpret the tool name prefix to infer such things. I guess I could consider try to add such logic directly into lld too, but the logic for dispatching between the different parts of lld (COFF/ELF/MachO/Wasm) is already quite complicated... For most purposes, `LLD_DEFAULT_LD_LLD_IS_MINGW` should be enough already.
